microbench/src/main/java/io/netty/buffer/ByteBufAccessBenchmark.java

```diff
@@ -0,0 +1,150 @@
+import io.netty.util.internal.PlatformDependent;
+
+@Warmup(iterations = 5, time = 1500, timeUnit = TimeUnit.MILLISECONDS)
+@Measurement(iterations = 10, time = 1500, timeUnit = TimeUnit.MILLISECONDS)
+@Fork(3)
+public class ByteBufAccessBenchmark extends AbstractMicrobenchmark {
+
+    static final class NioFacade extends WrappedByteBuf {
+        private final ByteBuffer byteBuffer;
+        NioFacade(ByteBuffer byteBuffer) {
+            super(Unpooled.EMPTY_BUFFER);
+            this.byteBuffer = byteBuffer;
+        }
+        @Override
+        public ByteBuf setLong(int index, long value) {
+            byteBuffer.putLong(index, value);
+            return this;
+        }
+        @Override
+        public long getLong(int index) {
+            return byteBuffer.getLong(index);
+        }
+        @Override
+        public byte readByte() {
+            return byteBuffer.get();
+        }
+        @Override
+        public ByteBuf touch() {
+            // hack since WrappedByteBuf.readerIndex(int) is final
+            byteBuffer.position(0);
+            return this;
+        }
+        @Override
+        public boolean release() {
+            PlatformDependent.freeDirectBuffer(byteBuffer);
+            return true;
+        }
+    }
+
+    public enum ByteBufType {
+        UNSAFE {
+            @Override
+            ByteBuf newBuffer() {
+                return new UnpooledUnsafeDirectByteBuf(
+                        UnpooledByteBufAllocator.DEFAULT, 64, 64).setIndex(0, 64);
+            }
+        },
+        UNSAFE_SLICE {
+            @Override
+            ByteBuf newBuffer() {
+                return UNSAFE.newBuffer().slice(16, 48);
+            }
+        },
+        HEAP {
+            @Override
+            ByteBuf newBuffer() {
+                return new UnpooledUnsafeHeapByteBuf(
+                        UnpooledByteBufAllocator.DEFAULT, 64, 64).setIndex(0,  64);
+            }
+        },
+        COMPOSITE {
+            @Override
+            ByteBuf newBuffer() {
+                return Unpooled.wrappedBuffer(UNSAFE.newBuffer(), HEAP.newBuffer());
+            }
+        },
+        NIO {
+            @Override
+            ByteBuf newBuffer() {
+                return new NioFacade(ByteBuffer.allocateDirect(64));
+            }
+        };
+        abstract ByteBuf newBuffer();
+    }
+
+    @Param
+    public ByteBufType bufferType;
+
+    private ByteBuf buffer;
+
+    @Setup
+    public void setup() {
+        buffer = bufferType.newBuffer();
+    }
+
+    @TearDown
+    public void tearDown() {
+        buffer.release();
+    }
+
+    @Benchmark
+    public long setGetLong() {
+        return buffer.setLong(0, 1).getLong(0);
+    }
+
+    @Benchmark
+    public ByteBuf setLong() {
+        return buffer.setLong(0, 1);
+    }
+
+    @Benchmark
+    public void readEightBytes(final Blackhole bh) {
+        buffer.readerIndex(0).touch();
+        bh.consume(buffer.readByte());
```

Comment:
It is what you want to measure? 
`ByteBuf::getByte(index)` is another interesting case: it won't have a data dependency that will cause bounds check, because the positions accessed are fixed (I need to verify it on the assembly).
Using `BlackHole::consume` between each `readByte` call will force (at least) a bounds check each time, because `readerIndex` updates cannot be moved; each consme include a load acquire barrier AFAIK.

Reply:
The numbers using the `readEightBytes` from above are actually very different (on the code of this PR).
Original `readEightBytes`:
```
Benchmark                              (bufferType)  (checkAccessible)  (checkBounds)   Mode  Cnt          Score          Error  Units
ByteBufAccessBenchmark.readEightBytes     COMPOSITE               true           true  thrpt   20   21234140.156 ±   540771.628  ops/s
ByteBufAccessBenchmark.readEightBytes     COMPOSITE               true          false  thrpt   20   25871522.124 ±   211593.779  ops/s
ByteBufAccessBenchmark.readEightBytes     COMPOSITE              false           true  thrpt   20   23694928.290 ±   288264.713  ops/s
ByteBufAccessBenchmark.readEightBytes     COMPOSITE              false          false  thrpt   20   27505818.147 ±   385209.348  ops/s
```
Modified `readEightBytes`:
```
Benchmark                              (bufferType)  (checkAccessible)  (checkBounds)   Mode  Cnt         Score         Error  Units
ByteBufAccessBenchmark.readEightBytes     COMPOSITE               true           true  thrpt   20  31660029.955 ± 1879564.146  ops/s
ByteBufAccessBenchmark.readEightBytes     COMPOSITE               true          false  thrpt   20  40709077.502 ± 2271395.159  ops/s
ByteBufAccessBenchmark.readEightBytes     COMPOSITE              false           true  thrpt   20  32216194.291 ± 1039418.921  ops/s
ByteBufAccessBenchmark.readEightBytes     COMPOSITE              false          false  thrpt   20  42907269.743 ±  819745.458  ops/s
```

Reply:
Thanks alot @franz1981, you're right that benchmark was flawed, I had actually been wondering why it didn't show better improvement in the context of the other results.

I guess an alternative is to "combine" the byte values somehow prior to consuming/returning although it may have some cost.

Reply:
Exactly, that's why I have proposed the version with multiple consume after a batch of read byte: ideally it would mimic what could happen while reading bytes from the buffer to accumulate them in another (single-threaded) data-structure

Reply:
If I may suggest replacing the 'read8' benchmark with a 'readBatch' which reads in a loop and "sums" the result somehow (e.g. `result += buf.readByte()`). The result can be sunk into a blackhole. The JVM has to read all the bytes to make the result and we want to evaluate the implications we have for batch reads IIUC. The loop will get unrolled which should end up being similar to the current case, and it would give us flexibility to test with different batch sizes.